docs: consistently mark 0 as valid anchor/reference

benbellick opened the pull request:
There are lots of places where the validity of anchor/reference 0 is not explicitly mentioned. This PR adds explicitly clarification in the proto comments and the site docs to clarify that anchor/reference 0 is valid. 

Closes #899

benbellick committed:
docs: consistently mark 0 as valid anchor/reference

benbellick commented on proto/substrait/algebra.proto:
```diff
@@ -1189,8 +1196,8 @@ message Expression {
-    // to a scalar function in the associated YAML file. Required; avoid
-    // using anchor/reference zero.
+    // to a scalar function in the associated YAML file.
+    // Required; 0 is a valid anchor/reference.
```
This is the only one I am not sure about... I find the language here a bit confusing. Is it saying the two separate things:
1. this field is required, and
2. you must not use anchor/reference 0
?

Or is it saying "it is required not to use anchor/reference 0"? 

I think it is the former. In which case there is nothing special about `ScalarFunction.function_reference` that prevents it from being 0.

vbarua replied:
I can confirm that the intent of this is:
1. It is required that this field be set.
2. 0 is a valid anchor / reference

benbellick commented on proto/substrait/type.proto:
```diff
@@ -10,6 +10,9 @@ option go_package = "github.com/substrait-io/substrait-protobuf/go/substraitpb";
 
+// Note: type_variation_reference fields within Type messages reference a
+// type_variation_anchor defined in the plan's extension declarations. The value
+// 0 represents the system-preferred variation and is a valid reference value.
```
Wasn't sure where a good place to put this was considering `type_variation_reference` shows up a ton of times.

yongchul replied:
For type variation anchor defined in the extension, it is better to not use `0` because it is *reserved* IMO. Perhaps, we can document this in the type extension proto message as well as type variation documentation?

vbarua commented:
I can provide some guidance on why the non-zero language exists for references. Effectively, if a `uint32` field in a protobuf message is not set explicilty, the value defaults to 0. This makes it difficult to distinguish unset references from those intentionally set to 0. When serializing protobuf for debugging/logging, by default `uint32` fields set to 0 are not output either which can make plans difficult read.

The stance of the project has generally been
```
0 is a valid anchor/reference, but prefer non-zero values for ergonomics
```

What do you think about including the explicit guidance to avoid zero values?

benbellick committed:
clarify that non-zero anchor preferred for ergonomics

vbarua reviewed (approved):
Thanks for updating this. Left on minor comment, but otherwise looks good to me.

benbellick commented on proto/substrait/type.proto:
```diff
@@ -10,6 +10,10 @@ option go_package = "github.com/substrait-io/substrait-protobuf/go/substraitpb";
+// Note: type_variation_reference fields within Type messages reference a
+// type_variation_anchor defined in the plan's extension declarations. The value
+// 0 represents the system-preferred variation and is a valid reference value,
+// but prefer non-zero values for ergonomics.
```
Good catch, what I wrote is incorrect!

benbellick committed:
drop mention of non-zero preference for `type_variation_reference`

benbellick committed:
Merge branch 'main' into benbellick/consistent-0-anchor

benbellick committed:
docs: clarify type_variation_anchor semantics and remove redundant reference comments

- Update type_variation_anchor comment to clarify 0 is reserved for system-preferred variation
- Reformat type.proto header comment for better readability
- Remove redundant anchor/reference comments from all *_reference fields since guidance on anchor definitions is sufficient
